### DISCUSSION TITLE
pb2nc generating many exactly 5000 m PBL heights

### DISCUSSION THREAD

**Discussion by gsketefian** [dtcenter/METplus#1140] [Statistical Computation] [answered]
Hi,
While doing verification of the FV3LAM's PBL heights, I encountered in the obs heights of exactly 5000 meters at many station locations (all at the same time).  In the attached figure for 2019053000, all the pink dots have a value of 5000.  Seems odd that this would be the case.  Has anyone else encountered this?  Thanks.
![PBL_height_obs_2019053000](https://user-images.githubusercontent.com/31046882/132061283-a5e49eab-3948-4a99-b427-0dbe5a3ffe55.png)

**Comment by hsoh-u** [comment]
The pb2nc resets PBL values to 5000 if it's bigger than 5000.
```
WARNING:
WARNING: Computed PBL (29794 from 29870) is too high, Reset to 5000  20210311_000000 ADPUPA 71082
WARNING:

29870 is the value from Fortran code
29794 is adjusted by header elevation
```

**Comment by JohnHalleyGotway** [comment, marked as answer]
Thanks @hsoh-u, here's the line of code that does that check with MAX_PBL = 5000 being the upper limit:
https://github.com/dtcenter/MET/blob/fb666bc3289e2bf49ccfc85e716ab812e10a7628/met/src/tools/other/pb2nc/pb2nc.cc#L3184

@gsketefian, as we discussed in an email thread with @jwolff-ncar, I pulled down some sample data and did more digging.

Using an input NDAS file from Hera (prepbufr.ndas.2019053000) and deriving PBL in PB2NC results in 54 of 136 PBL values being computed as 5000. Here's a histogram showing this result:
<img width="494" alt="Screen Shot 2021-09-03 at 2 59 19 PM" src="https://user-images.githubusercontent.com/21087144/132064800-8c1f6308-38d4-4027-aab6-fb93aed70eaa.png">

Running pb2nc at verbosity level 7 (-v 7) produces lots of useful info about the derivation of PBL in this 
[run_pb2nc.log](https://github.com/dtcenter/METplus/files/7108474/run_pb2nc.log) file.

PBL derivation results in two types of warning messages:
1. When PBL is less than the station elevation:
```
WARNING: Not saved because the computed PBL (-2.52435e-29) is less than the station elevation (601). 20190530_000000 ADPUPA 71908
```
2. When PBL is greater than the prescribed upper limit:
```
WARNING: Computed PBL (1.58495e+29 from 1.58495e+29) is too high, Reset to 5000  20190530_000000 ADPUPA 71802
```
While values too low are discarded, values too high are just reset to 5000. I recommend that we modify this logic and also discard the large values. I tested with a modified version of PB2NC and here's a much more reasonable histogram of computed PBL values:
<img width="491" alt="Screen Shot 2021-09-03 at 11 54 27 AM" src="https://user-images.githubusercontent.com/21087144/132065228-be00890b-4f3d-4ea1-9cb3-0cb8576fc237.png">

However, I'd like @PerryShafran-NOAA to weigh in. The PBL derivation logic was ported over from NOAA/EMC verification code. @PerryShafran-NOAA is resetting large values of PBL to 5000 a useful feature? Or should we instead just discard those obs?

Note that the vast majority of the values that are reset are ridiculous and certainly not meaningful:
```
WARNING: Computed PBL (29236 from 29566) is too high, Reset to 5000  20190530_000000 ADPUPA 72632
WARNING: Computed PBL (1.58495e+29 from 1.58495e+29) is too high, Reset to 5000  20190530_000000 ADPUPA 71845
WARNING: Computed PBL (8.59202e+09 from 8.59202e+09) is too high, Reset to 5000  20190530_000000 ADPUPA 71945
WARNING: Computed PBL (1.58495e+29 from 1.58495e+29) is too high, Reset to 5000  20190530_000000 ADPUPA 72440
WARNING: Computed PBL (31986 from 32309) is too high, Reset to 5000  20190530_000000 ADPUPA 72426
WARNING: Computed PBL (1.58495e+29 from 1.58495e+29) is too high, Reset to 5000  20190530_000000 ADPUPA 72747
WARNING: Computed PBL (3.69025e+19 from 3.69025e+19) is too high, Reset to 5000  20190530_000000 ADPUPA 72249
WARNING: Computed PBL (8.59202e+09 from 8.59202e+09) is too high, Reset to 5000  20190530_000000 ADPUPA 74455
WARNING: Computed PBL (1.58495e+29 from 1.58495e+29) is too high, Reset to 5000  20190530_000000 ADPUPA 71802
WARNING: Computed PBL (8.59202e+09 from 8.59202e+09) is too high, Reset to 5000  20190530_000000 ADPUPA 70231
```
Does this indicate some fault in the PBL derivation algorithm? Or is this expected behavior?

@gsketefian while we sort out the details, there is an easy workaround you can use. In your Point-Stat configuration file, you can censor the PBL observations using:
```
obs = {
   field = [
   ...
      { name = "PBL"; level = "L0"; censor_thresh = [ >=5000 ]; censor_val = [ -9999 ]; }
   ...
```
This replaces all instances of 5000 with MET's internal bad data value (-9999). And those values will no longer be included in the verification. Unfortunately this does require rerunning Point-Stat.

**Reply by jwolff-ncar** [reply]
Thank you @JohnHalleyGotway and @hsoh-u for following up. Appreciate the work-around while this behavior is discussed further with @PerryShafran-NOAA to decide on a final fix.

**Reply by gsketefian** [reply]
Yes, thanks @JohnHalleyGotway and @hsoh-u for the answers.  Let's see what @PerryShafran-NOAA says and go from there.

**Reply by PerryShafran-NOAA** [reply]
I think I might want to check with the team to see if there should be a max level of PBL height.  5000 m is a real physical value, and in some very stable regions, PBL height can get even above that.  I'm not entirely sure if there is a max value in the models, but I can also check that out.  

I'm not certain of why there was a 5000 m max in the VSDB code, I wasn't the author of that, I just took that as face value.  Let me confirm to see if there is a current max and what the code should do with those.

**Reply by JohnHalleyGotway** [reply]
Thanks for taking a look @PerryShafran-NOAA. Please let me know what you find.

@hsoh-u from a software perspective, I'm concerned about some pretty clearly bad data values being computed. Listed below is some debug level 8 log output from PB2NC. For station number 72274, the computed PBL value, 3.68935e+19, is rejected because its greater than 5000. But we see that that value comes from the first input level... and is clearly not physical. I wonder where these ridiculous values are coming from, and wonder if there's any additional checks we should add to interpret them as bad data? 
```
DEBUG 8:
DEBUG 8: compute_pbl()     input to calpbl_: 0   0.00000 0.00000 36893488147419103232.00000 36893488147419103232.00000 -2.29688 1.30000
DEBUG 8: compute_pbl()     input to calpbl_: 1   2810.00000 220.15364 0.00001 24384.00000 -7.00000 3.30000
DEBUG 8: compute_pbl()     input to calpbl_: 2   3000.00000 219.25000 0.00001 23960.00000 -9.50000 2.50000
DEBUG 8: compute_pbl()     input to calpbl_: 3   3090.00000 218.82179 0.00001 23774.00000 -9.30000 -4.30000
DEBUG 8: compute_pbl()     input to calpbl_: 4   3920.00000 215.37520 0.00000 22250.00000 -1.90000 0.90000
DEBUG 8: compute_pbl()     input to calpbl_: 5   4530.00000 213.28004 0.00000 21336.00000 -7.70000 0.00000
DEBUG 8: compute_pbl()     input to calpbl_: 6   5000.00000 211.85001 0.00000 20720.00000 -0.40000 1.40000
DEBUG 8: compute_pbl()     input to calpbl_: 7   6390.00000 210.10034 0.00000 19202.00000 0.10000 0.50000
DEBUG 8: compute_pbl()     input to calpbl_: 8   6710.00000 209.75180 0.00000 18898.00000 3.80000 5.50000
DEBUG 8: compute_pbl()     input to calpbl_: 9   7000.00000 209.45000 0.00000 18640.00000 1.00000 5.60000
DEBUG 8: compute_pbl()     input to calpbl_: 10  7050.00000 209.44202 0.00000 18593.00000 -1.30000 4.90000
DEBUG 8: compute_pbl()     input to calpbl_: 11  7419.99951 209.38466 0.00000 18288.00000 -0.50000 1.40000
DEBUG 8: compute_pbl()     input to calpbl_: 12  7800.00000 209.32864 0.00000 17983.00000 4.00000 4.00000
DEBUG 8: compute_pbl()     input to calpbl_: 13  8600.00000 209.21915 0.00000 17374.00000 10.40000 2.80000
DEBUG 8: compute_pbl()     input to calpbl_: 14  9040.00000 209.16319 0.00000 17069.00000 12.50000 7.20000
DEBUG 8: compute_pbl()     input to calpbl_: 15  10000.00000 209.05000 0.00000 16450.00000 10.60000 1.90000
DEBUG 8: compute_pbl()     input to calpbl_: 16  12190.00000 210.22217 0.00000 15240.00000 21.40000 12.30000
DEBUG 8: compute_pbl()     input to calpbl_: 17  12810.00098 210.51582 0.00000 14935.00000 13.50000 9.50000
DEBUG 8: compute_pbl()     input to calpbl_: 18  13470.00000 210.81319 0.00000 14630.00000 17.40000 14.60000
DEBUG 8: compute_pbl()     input to calpbl_: 19  15000.00000 211.45000 0.00000 13960.00000 23.20000 8.40000
DEBUG 8: compute_pbl()     input to calpbl_: 20  15610.00098 211.14314 0.00000 13716.00000 33.30000 8.90000
DEBUG 8: compute_pbl()     input to calpbl_: 21  15800.00000 211.05000 0.00000 13647.06836 32.40000 8.70000
DEBUG 8: compute_pbl()     input to calpbl_: 22  18970.00000 216.63501 0.00001 12497.00000 27.10000 12.60000
DEBUG 8: compute_pbl()     input to calpbl_: 23  20000.00000 218.25000 0.00001 12160.00000 31.50000 11.50000
DEBUG 8: compute_pbl()     input to calpbl_: 24  25000.00000 227.45000 0.00004 10700.00000 35.30000 9.50000
DEBUG 8: compute_pbl()     input to calpbl_: 25  28790.00000 233.95325 0.00005 9754.00000 35.90000 16.70000
DEBUG 8: compute_pbl()     input to calpbl_: 26  30000.00000 235.85001 0.00005 9470.00000 35.30000 12.90000
DEBUG 8: compute_pbl()     input to calpbl_: 27  31440.00000 238.45752 0.00006 9144.00000 33.30000 8.90000
DEBUG 8: compute_pbl()     input to calpbl_: 28  38930.00000 250.34196 0.00009 7620.00000 28.20000 -2.50000
DEBUG 8: compute_pbl()     input to calpbl_: 29  40000.00000 251.85001 0.00009 7420.00000 27.30000 0.00000
DEBUG 8: compute_pbl()     input to calpbl_: 30  44020.00000 255.36911 0.00025 6706.00000 16.00000 5.80000
DEBUG 8: compute_pbl()     input to calpbl_: 31  47770.00000 258.37335 0.00039 6096.00000 14.80000 1.30000
DEBUG 8: compute_pbl()     input to calpbl_: 32  50000.00000 260.04999 0.00046 5750.00000 10.80000 0.00000
DEBUG 8: compute_pbl()     input to calpbl_: 33  51770.00000 261.92136 0.00068 5486.00000 8.70000 -1.50000
DEBUG 8: compute_pbl()     input to calpbl_: 34  56020.00000 266.16553 0.00117 4877.00000 4.80000 4.00000
DEBUG 8: compute_pbl()     input to calpbl_: 35  58240.00391 268.25616 0.00141 4572.00000 4.90000 2.80000
DEBUG 8: compute_pbl()     input to calpbl_: 36  60530.00000 270.33078 0.00164 4267.00000 4.00000 -1.10000
DEBUG 8: compute_pbl()     input to calpbl_: 37  65320.00000 274.42761 0.00212 3658.00000 1.80000 2.50000
DEBUG 8: compute_pbl()     input to calpbl_: 38  70000.00000 278.14999 0.00254 3098.00000 2.50000 0.70000
DEBUG 8: compute_pbl()     input to calpbl_: 39  73090.00000 281.62067 0.00255 2743.00000 2.00000 0.50000
DEBUG 8: compute_pbl()     input to calpbl_: 40  75830.00000 284.57767 0.00255 2438.00000 2.10000 -1.50000
DEBUG 8: compute_pbl()     input to calpbl_: 41  78640.00000 287.50125 0.00256 2134.00000 1.40000 -3.80000
DEBUG 8: compute_pbl()     input to calpbl_: 42  81520.00000 290.39117 0.00256 1829.00000 0.40000 -4.60000
DEBUG 8: compute_pbl()     input to calpbl_: 43  85000.00000 293.75000 0.00257 1471.00000 2.50000 -4.40000
DEBUG 8: compute_pbl()     input to calpbl_: 44  87510.00000 296.64627 0.00251 1219.00000 3.30000 -3.30000
DEBUG 8: compute_pbl()     input to calpbl_: 45  90610.00000 300.11087 0.00244 914.00000 6.10000 -2.80000
DEBUG 8: compute_pbl()     input to calpbl_: 46  92300.00000 301.95001 0.00240 751.00000 -9999.00000 -9999.00000
DEBUG 7:  hpbl: 3.68935e+19, obs_arr[4]: 3.68935e+19   lat: 32.23, lon: -110.96, elv: 751 20190530_000000 ADPUPA 72274
DEBUG 7:
WARNING:
WARNING: Skipping out of range PBL (3.68935e+19 > 5000) from PBL value (3.68935e+19) minus station elevation (751) for 20190530_000000 ADPUPA 72274.
WARNING:
```

**Comment by JeffBeck-NOAA** [comment]
I agree that 5000 m is too low to safely claim anything at that height or higher is invalid.  I'd feel comfortable with ~7000 m, but again, this comes back to how the astronomically high PBL values are getting created in the first place.  @JohnHalleyGotway, does MET need to be rebuilt to change MAX_PBL, or can that be set at run time?

**Reply by JohnHalleyGotway** [reply]
@JeffBeck-NOAA the MAX_PBL value is hard-coded on this line:
https://github.com/dtcenter/MET/blob/fb666bc3289e2bf49ccfc85e716ab812e10a7628/met/src/tools/other/pb2nc/pb2nc.cc#L175

Changing it from 5000 to some other value, like 7000, would require that MET be recompiled. At this point, I see three relevant questions that need to be addressed:

(1) @hsoh-u and @PerryShafran-NOAA: How can we detect and handle bad input values (like level 0 height = 3.68935e+19, in the example above)? Am I right in assuming the resulting PBL should be bad data rather than 5000?

(2) @JeffBeck-NOAA and @PerryShafran-NOAA: Should the max PBL value (currently 5000) be modified? And should values above the max be reset to that max value (current behavior) or reset to bad data?

(3) Should we make the definition of that max PBL a configurable option instead of hard coding it? If so, consider using the existing censor_thresh and censor_val options to define how to censor the output observations. Something like this?
```
obs_censor = [
   { var  = "PBL"; censor_thresh = [ gt7000; ]; censor_val = [ 7000 ]; }
];
```
Defining it in this general way would give users a lot of flexibility  in censoring the observation data. If this is worthwhile that'd require a new GitHub issue.

**Reply by JeffBeck-NOAA** [reply]
@JohnHalleyGotway, regarding 2) Yes, 5000 m should not be the max PBL value.  If bad values are always 3e19 or something huge, then setting a max PBL of, say, 10000m would be fine, and would capture all valid values of PBL.  Then values above that should be thrown out.  5000 m is just too low of a threshold.  Regarding 3), I like the idea of allowing users flexibility to set their PBL height max; however, if the difference between a good and "bad" PBL height value is huge, then there's no need to do this.

**Reply by PerryShafran-NOAA** [reply]
I am asking the verification team thoughts about max PBL height.  We are waiting for the relevant people to return from vacation to help give us an answer.  But I think that there are realistic PBL heights above 5000 in certain regimes.

We would also be limited by a max value in the UPP, which I'm not entirely sure if there is a threshold there.  We will have to conform to what would be in our UPP has in order to be consistent.

**Comment by gsketefian** [comment]
@PerryShafran-NOAA @JohnHalleyGotway @hsoh-u We (I, @jwolff-ncar, and @JeffBeck-NOAA, and others who are part of the DTC's R2O project) are looking to rerun our verification tasks with a modification to account for the 5000m HPBL bug.  Since we need to do that pretty soon, we are trying to decide on how to proceed.  The options are:

1. Just throw out the 5000m PBL height values.
2. Wait until someone who is familiar with the HPBL code determines why the code is sometimes producing the unrealistic values (on the order of e+29).  If these unrealistic values are due to something that can be easily fixed (e.g. bad level 0 values, as @JohnHalleyGotway pointed out), it would be nice to make the fix and include those observations in our analysis.

@PerryShafran-NOAA, just wondering if you have an estimate on when the relevant people will be able to take a look at the code.  Thanks.

**Reply by PerryShafran-NOAA** [reply]
Hi, there, it seems to me that our validation folks seem to agree that 5000 m is too low a cap, but I'm trying to figure out whether they want some sort of cap at all, or just to throw out the unrealistic values.  I'll try to get an answer ASAP.

For the unrealistic values, it would be useful to plot out all the PBL components at every level and find the unrealistic value in there.  I would be willing to assist.  

It should be noted that EMC does not agree that we should allow users to select their own level to throw out PBL data.

**Reply by PerryShafran-NOAA** [reply]
Confirmed:  set the max level to 10,000 m, and that should capture the physical PBL heights.  Anything about that should be flagged and thrown out.

**Reply by JohnHalleyGotway** [reply]
@PerryShafran-NOAA thanks for the direction. Two questions for you...

(1) @hsoh-u noticed in this [example](https://github.com/dtcenter/METplus/discussions/1140#discussioncomment-1292156), that the first level has a pressure of 0.00000. We assume that we should treat any input levels with a pressure value of 0 as being bad data. Can you confirm that's correct? And if so, @hsoh-u please write up a GitHub issue for this.

(2) For the maximum value of 10,000, and resetting anything above that to bad data, do you think we should hard-code that upper limit? Or would you prefer that we handle it via the PB2NC configuration file by adding a new option for censoring point obsrevations values:
```
obs_censor = [
   { var  = "PBL"; censor_thresh = [ gt10000; ]; censor_val = [ -9999 ]; }
];
```
I suspect that we're likely to encounter issues similar to this in the future.
Once I hear back from you, I can write up a new GitHub issue to capture this.

**Reply by PerryShafran-NOAA** [reply]
Hi, @JohnHalleyGotway ,

1)  Yes, that assumption should be correct.  If something has a non-physical value, consider that bad data.  
2)  I think it should be hard coded.  We do not want to have users to have the option of selecting that upper value.  

Thanks!

Perry

**Reply by JohnHalleyGotway** [reply]
Thanks for clarifying Perry. I migrated this direction over to a new METplus issue dtcenter/METplus#1913. @hsoh-u will work on this for MET version 10.1.0. I also assigned Perry to that issue for him to test the result.

And I'll select one of the responses above as the "Answer" to this discussion.

**Comment by gsketefian** [comment]
@PerryShafran-NOAA @JohnHalleyGotway @hsoh-u Couple of questions:
1. Does the 0-pressure bug show up in the output in the [example](https://github.com/dtcenter/METplus/discussions/1140#discussioncomment-1292156) above?  The first column of numbers seems to be the pressure in Pa.  Level 0 (which seems to be the top) seems to have a pressure of 0 (which sort of makes sense), and level 46 (which seems to be the surface layer) has a pressure of 92300 Pa, which is reasonable.  I do not see a zero pressure at the surface.
2. Assuming I misunderstood and there is a 0-pressure bug somewhere, do we know that all the bad 5000m values in the [plot](https://github.com/dtcenter/METplus/discussions/1140#discussion-3554764) above (the violet dots) are due to that bug?  It seems improbable that so many of the stations (eyeballing, maybe 60% of them) would all have bad pressures at the same time.  

Would appreciate any feedback.  Thanks.

**Reply by PerryShafran-NOAA** [reply]
It is potentially likely that those PBL heights over 5000 m are real, as it is possible to have a PBL value more than 5000 m.  It would be good to know what those PBL values those are that were reset to 5000 m.  

Also, our radiosondes generally do not measure all the way up to 0 hPa, so if that is the pressure, then it's likely a bad observation.  Looking at that level, none of the observations in that level are reasonable.  Therefore, if it's just one level, you could throw away the level of that observation and calculate the PBL using all the other levels with reasonable data.

**Comment by JohnHalleyGotway** [comment]
Please note that this discussion was moved over to a MET development issue:
https://github.com/dtcenter/MET/issues/1913
That work was completed and the changes are included in the met-10.1.0-beta3 development release. For this reason, I'm locking this discussion to basically resolve it. Please create a new discussion if future questions or issues arise.
